incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSessionFactory.java

```diff
@@ -1026,49 +1106,95 @@ private void onKafkaData(
                     kafkaDataEx != null && kafkaDataEx.kind() == KafkaDataExFW.KIND_MERGED ? kafkaDataEx.merged() : null;
                 final KafkaKeyFW key = kafkaMergedDataEx != null ? kafkaMergedDataEx.key() : null;
 
-                fetchWill:
+                reactToSignal:
                 if (key != null)
                 {
                     if (payload == null)
                     {
+                        final OctetsFW typeHeaderNameOctets =
+                            new OctetsFW().wrap(TYPE_HEADER_NAME.value(), 0, TYPE_HEADER_NAME.length());
+                        final String type = kafkaMergedDataEx.headers().matchFirst(h -> h.name().equals(typeHeaderNameOctets))
+                            .value().get((b, o, m) -> b.getStringWithoutLengthUtf8(o, m - o));
+
+                        final String keyPostfix = type.equals(WILL_SIGNAL_NAME.asString()) ?
+                            WILL_SIGNAL_KEY_POSTFIX : EXPIRY_SIGNAL_KEY_POSTFIX;
```

Comment:
`type` can be `OctetsFW` and `WILL_SIGNAL_NAME` can be preallocated as `OctetsFW` so that .equals still works, no need for `.asString()`.